File: arsceneview/src/main/java/io/github/sceneview/ar/rerun/RerunWireFormat.kt
package io.github.sceneview.ar.rerun

import com.google.ar.core.Anchor
import com.google.ar.core.HitResult
import com.google.ar.core.Plane
import com.google.ar.core.PointCloud
import com.google.ar.core.Pose

/**
 * JSON-lines wire format shared between the Kotlin and Swift Rerun bridges.
 *
 * Every event is one JSON object on a single line, terminated by `\n`, and
 * consumed by a Python sidecar that re-logs it via the `rerun-sdk` into the
 * Rerun viewer. The same wire format is emitted on Android (ARCore) and iOS
 * (ARKit) so a single sidecar script can handle both platforms.
 *
 * Format:
 * ```
 * {"t": 123456789, "type": "camera_pose", "entity": "world/camera", "translation": [x,y,z], "quaternion": [x,y,z,w]}
 * {"t": 123456789, "type": "plane", "entity": "world/planes/<id>", "polygon": [[x,y,z], ...], "kind": "horizontal"}
 * {"t": 123456789, "type": "point_cloud", "entity": "world/points", "positions": [[x,y,z], ...], "confidences": [f, ...]}
 * {"t": 123456789, "type": "anchor", "entity": "world/anchors/<id>", "translation": [x,y,z], "quaternion": [x,y,z,w]}
 * {"t": 123456789, "type": "hit_result", "entity": "world/hits/<id>", "translation": [x,y,z], "distance": f}
 * ```
 *
 * This module is pure: no I/O, no threading, no Android framework. It's a
 * `String`-in-`String`-out serializer that can be unit-tested on plain JVM.
 * All the network plumbing lives in [RerunBridge].
 *
 * JSON is hand-rolled via [StringBuilder] so the runtime has zero new
 * dependencies — `kotlinx.serialization` is not on the `arsceneview`
 * classpath and pulling it in for a debug-only logger would bloat every
 * SceneView app by ~120 kB.
 */
internal object RerunWireFormat {

    /** Camera pose event — one per frame. Entity path defaults to `world/camera`. */
    fun cameraPose(
        timestampNanos: Long,
        pose: Pose,
        entity: String = "world/camera",
    ): String = cameraPoseJson(
        timestampNanos = timestampNanos,
        tx = pose.tx(), ty = pose.ty(), tz = pose.tz(),
        qx = pose.qx(), qy = pose.qy(), qz = pose.qz(), qw = pose.qw(),
        entity = entity,
    )

    /**
     * Testable overload that takes primitives instead of a [Pose].
     *
     * ARCore's [Pose] class is native and final, so mocking it in a pure
     * JVM unit test requires mockito-inline or equivalent. This function
     * lets the golden-JSON tests assert exact output without touching
     * ARCore at all.
     */
    internal fun cameraPoseJson(
        timestampNanos: Long,
        tx: Float, ty: Float, tz: Float,
        qx: Float, qy: Float, qz: Float, qw: Float,
        entity: String = "world/camera",
    ): String = buildString(192) {
        append('{')
        appendCommonHeader(timestampNanos, "camera_pose", entity)
        append(",\"translation\":[")
        appendFloat(tx); append(','); appendFloat(ty); append(','); appendFloat(tz)
        append("],\"quaternion\":[")
        appendFloat(qx); append(','); appendFloat(qy); append(','); appendFloat(qz); append(','); appendFloat(qw)
        append("]}\n")
    }

    /** Testable overload for anchors — takes the pose as primitives. */
    internal fun anchorJson(
        timestampNanos: Long,
        id: Int,
        tx: Float, ty: Float, tz: Float,
        qx: Float, qy: Float, qz: Float, qw: Float,
    ): String = buildString(192) {
        append('{')
        appendCommonHeader(timestampNanos, "anchor", "world/anchors/$id")
        append(",\"translation\":[")
        appendFloat(tx); append(','); appendFloat(ty); append(','); appendFloat(tz)
        append("],\"quaternion\":[")
        appendFloat(qx); append(','); appendFloat(qy); append(','); appendFloat(qz); append(','); appendFloat(qw)
        append("]}\n")
    }

    /** Testable overload for hit results. */
    internal fun hitResultJson(
        timestampNanos: Long,
        id: Int,
        tx: Float, ty: Float, tz: Float,
        distance: Float,
    ): String = buildString(160) {
        append('{')
        appendCommonHeader(timestampNanos, "hit_result", "world/hits/$id")
        append(",\"translation\":[")
        appendFloat(tx); append(','); appendFloat(ty); append(','); appendFloat(tz)
        append("],\"distance\":")
        appendFloat(distance)
        append("}\n")
    }

    /**
     * Testable overload for point clouds — takes the raw `[x,y,z,confidence]`
     * packed FloatArray that ARCore's `PointCloud.points` buffer yields.
     */
    internal fun pointCloudJson(
        timestampNanos: Long,
        packed: FloatArray,
        entity: String = "world/points",
    ): String {
        val n = packed.size / 4
        return buildString(64 + n * 48) {
            append('{')
            appendCommonHeader(timestampNanos, "point_cloud", entity)
            append(",\"positions\":[")
            for (i in 0 until n) {
                if (i > 0) append(',')
                append('[')
                appendFloat(packed[i * 4]); append(',')
                appendFloat(packed[i * 4 + 1]); append(',')
                appendFloat(packed[i * 4 + 2])
                append(']')
            }
            append("],\"confidences\":[")
            for (i in 0 until n) {
                if (i > 0) append(',')
                appendFloat(packed[i * 4 + 3])
            }
            append("]}\n")
        }
    }

    /**
     * Testable overload for planes — takes the already-world-space polygon
     * and plane kind so tests don't need a real ARCore Plane instance.
     */
    internal fun planeJson(
        timestampNanos: Long,
        id: Int,
        kind: String,
        worldPolygon: List<FloatArray>,
    ): String = buildString(256 + worldPolygon.size * 48) {
        append('{')
        appendCommonHeader(timestampNanos, "plane", "world/planes/$id")
        append(",\"kind\":\"")
        append(kind)
        append("\",\"polygon\":[")
        for (idx in worldPolygon.indices) {
            if (idx > 0) append(',')
            val p = worldPolygon[idx]
            append('[')
            appendFloat(p[0]); append(','); appendFloat(p[1]); append(','); appendFloat(p[2])
            append(']')
        }
        append("]}\n")
    }

    /**
     * Plane event — one per detected ARCore [Plane].
     *
     * The plane polygon is emitted as world-space 3D points (ARCore's local
     * 2D polygon is lifted into world coordinates via the plane's center pose
     * so the Python sidecar can draw a closed LineStrip3D directly).
     */
    fun plane(
        timestampNanos: Long,
        plane: Plane,
    ): String {
        // Stable id: ARCore Plane identity is object identity; use System.identityHashCode
        // so the Python sidecar can consistently attach the same entity path across frames.
        val id = System.identityHashCode(plane)
        // Keep a defensive fallback (silenced) so a future ARCore Plane.Type
        // addition doesn't crash the bridge — it just shows up as "unknown".
        @Suppress("REDUNDANT_ELSE_IN_WHEN")
        val kind = when (plane.type) {
            Plane.Type.HORIZONTAL_UPWARD_FACING -> "horizontal_upward"
            Plane.Type.HORIZONTAL_DOWNWARD_FACING -> "horizontal_downward"
            Plane.Type.VERTICAL -> "vertical"
            else -> "unknown"
        }

        // Lift the local 2D polygon [x0,z0, x1,z1, ...] into world-space 3D points.
        val localPoly = plane.polygon
        val centerPose = plane.centerPose
        val worldPoints = FloatArray(3)
        val worldPoly = ArrayList<FloatArray>(localPoly.remaining() / 2)
        val copy = FloatArray(localPoly.remaining())
        localPoly.get(copy)
        var i = 0
        while (i < copy.size - 1) {
            val local = floatArrayOf(copy[i], 0f, copy[i + 1])
            centerPose.transformPoint(local, 0, worldPoints, 0)
            worldPoly.add(worldPoints.copyOf())
            i += 2
        }

        return planeJson(timestampNanos, id, kind, worldPoly)
    }

    /**
     * Point cloud event — one per frame.
     *
     * ARCore's [PointCloud.points] is a packed `[x, y, z, confidence]` float
     * buffer. We split it into `positions[[x,y,z], ...]` and
     * `confidences[f, ...]` for easier consumption in Python.
     */
    fun pointCloud(
        timestampNanos: Long,
        cloud: PointCloud,
        entity: String = "world/points",
    ): String {
        val buf = cloud.points
        // Defensive copy so we don't mutate ARCore's read-position.
        val data = FloatArray(buf.remaining())
        buf.asReadOnlyBuffer().get(data)
        return pointCloudJson(timestampNanos, data, entity)
    }

    /** Anchor event — one per user-placed anchor. */
    fun anchor(
        timestampNanos: Long,
        anchor: Anchor,
    ): String {
        val id = System.identityHashCode(anchor)
        val p = anchor.pose
        return anchorJson(
            timestampNanos, id,
            p.tx(), p.ty(), p.tz(),
            p.qx(), p.qy(), p.qz(), p.qw(),
        )
    }

    /** Hit result event — raycast intersection. */
    fun hitResult(
        timestampNanos: Long,
        hit: HitResult,
    ): String {
        val id = System.identityHashCode(hit)
        val p = hit.hitPose
        return hitResultJson(
            timestampNanos, id,
            p.tx(), p.ty(), p.tz(),
            hit.distance,
        )
    }

    // ── Helpers ────────────────────────────────────────────────────────────

    private fun StringBuilder.appendCommonHeader(t: Long, type: String, entity: String) {
        append("\"t\":")
        append(t)
        append(",\"type\":\"")
        append(type)
        append("\",\"entity\":\"")
        appendEscaped(entity)
        append('"')
    }

    /**
     * Writes a float as plain JSON number. Non-finite values (NaN/Infinity)
     * would break the JSON line; we emit `0` instead so the line stays parseable.
     * The Python sidecar treats tiny outliers as noise so clamping is safe.
     */
    private fun StringBuilder.appendFloat(f: Float) {
        if (!f.isFinite()) {
            append('0')
            return
        }
        append(f)
    }

    /**
     * Minimal JSON string escaper. Entity paths are developer-supplied so
     * they may legitimately contain `/` — but never a `"` or control char
     * under our conventions. Still, we escape defensively to avoid a single
     * rogue id corrupting the wire format.
     */
    private fun StringBuilder.appendEscaped(s: String) {
        for (c in s) {
            when {
                c == '"' -> append("\\\"")
                c == '\\' -> append("\\\\")
                c == '\n' -> append("\\n")
                c == '\r' -> append("\\r")
                c == '\t' -> append("\\t")
                c.code < 0x20 -> append(String.format("\\u%04x", c.code))
                else -> append(c)
            }
        }
    }
}

File: SceneViewSwift/Sources/SceneViewSwift/Rerun/RerunWireFormat.swift
import Foundation
#if os(iOS)
import ARKit
#endif

/// JSON-lines wire format shared between the Kotlin and Swift Rerun bridges.
///
/// Every event is one JSON object on a single line, terminated by `\n`, and
/// consumed by a Python sidecar that re-logs it via the `rerun-sdk` into the
/// Rerun viewer. The same wire format is emitted on Android (ARCore) and
/// iOS (ARKit), so a single sidecar script handles both platforms.
///
/// Format:
/// ```
/// {"t": 123456789, "type": "camera_pose", "entity": "world/camera", "translation": [x,y,z], "quaternion": [x,y,z,w]}
/// {"t": 123456789, "type": "plane", "entity": "world/planes/<id>", "polygon": [[x,y,z], ...], "kind": "horizontal"}
/// {"t": 123456789, "type": "point_cloud", "entity": "world/points", "positions": [[x,y,z], ...], "confidences": [f, ...]}
/// {"t": 123456789, "type": "anchor", "entity": "world/anchors/<id>", "translation": [x,y,z], "quaternion": [x,y,z,w]}
/// {"t": 123456789, "type": "hit_result", "entity": "world/hits/<id>", "translation": [x,y,z], "distance": f}
/// ```
///
/// This module is pure: no I/O, no threading, no ARKit instances needed by
/// the testable `*Json` overloads — the string-in / string-out nature lets
/// the unit tests assert exact golden JSON output on plain Swift.
///
/// JSON is hand-rolled instead of using `JSONEncoder` so Kotlin and Swift
/// emit byte-identical output for the same input — the Python sidecar
/// doesn't care about whitespace but the cross-platform golden tests do.
public enum RerunWireFormat {

    // MARK: - Testable overloads (primitives only)

    /// Camera pose event — takes the pose as primitives so tests don't need
    /// a real ARKit frame.
    public static func cameraPoseJson(
        timestampNanos: Int64,
        tx: Float, ty: Float, tz: Float,
        qx: Float, qy: Float, qz: Float, qw: Float,
        entity: String = "world/camera"
    ) -> String {
        var s = ""
        s.reserveCapacity(192)
        s.append("{")
        appendCommonHeader(&s, timestampNanos: timestampNanos, type: "camera_pose", entity: entity)
        s.append(",\"translation\":[")
        appendFloat(&s, tx); s.append(",")
        appendFloat(&s, ty); s.append(",")
        appendFloat(&s, tz)
        s.append("],\"quaternion\":[")
        appendFloat(&s, qx); s.append(",")
        appendFloat(&s, qy); s.append(",")
        appendFloat(&s, qz); s.append(",")
        appendFloat(&s, qw)
        s.append("]}\n")
        return s
    }

    /// Anchor event — takes pose as primitives.
    public static func anchorJson(
        timestampNanos: Int64,
        id: Int,
        tx: Float, ty: Float, tz: Float,
        qx: Float, qy: Float, qz: Float, qw: Float
    ) -> String {
        var s = ""
        s.reserveCapacity(192)
        s.append("{")
        appendCommonHeader(&s, timestampNanos: timestampNanos, type: "anchor", entity: "world/anchors/\(id)")
        s.append(",\"translation\":[")
        appendFloat(&s, tx); s.append(",")
        appendFloat(&s, ty); s.append(",")
        appendFloat(&s, tz)
        s.append("],\"quaternion\":[")
        appendFloat(&s, qx); s.append(",")
        appendFloat(&s, qy); s.append(",")
        appendFloat(&s, qz); s.append(",")
        appendFloat(&s, qw)
        s.append("]}\n")
        return s
    }

    /// Hit result event — takes position + distance.
    public static func hitResultJson(
        timestampNanos: Int64,
        id: Int,
        tx: Float, ty: Float, tz: Float,
        distance: Float
    ) -> String {
        var s = ""
        s.reserveCapacity(160)
        s.append("{")
        appendCommonHeader(&s, timestampNanos: timestampNanos, type: "hit_result", entity: "world/hits/\(id)")
        s.append(",\"translation\":[")
        appendFloat(&s, tx); s.append(",")
        appendFloat(&s, ty); s.append(",")
        appendFloat(&s, tz)
        s.append("],\"distance\":")
        appendFloat(&s, distance)
        s.append("}\n")
        return s
    }

    /// Point cloud event — takes already-split positions and confidences.
    /// Positions is a flat `[x0, y0, z0, x1, y1, z1, ...]` buffer; each 3
    /// floats form one point. Confidences is parallel, one per point.
    public static func pointCloudJson(
        timestampNanos: Int64,
        positions: [Float],
        confidences: [Float] = [],
        entity: String = "world/points"
    ) -> String {
        let n = positions.count / 3
        var s = ""
        s.reserveCapacity(64 + n * 48)
        s.append("{")
        appendCommonHeader(&s, timestampNanos: timestampNanos, type: "point_cloud", entity: entity)
        s.append(",\"positions\":[")
        for i in 0..<n {
            if i > 0 { s.append(",") }
            s.append("[")
            appendFloat(&s, positions[i * 3]); s.append(",")
            appendFloat(&s, positions[i * 3 + 1]); s.append(",")
            appendFloat(&s, positions[i * 3 + 2])
            s.append("]")
        }
        s.append("],\"confidences\":[")
        for i in 0..<confidences.count {
            if i > 0 { s.append(",") }
            appendFloat(&s, confidences[i])
        }
        s.append("]}\n")
        return s
    }

    /// Plane event — takes a world-space polygon as a list of 3-float
    /// arrays. ARKit gives us a `simd_float3` array from the plane anchor
    /// geometry; the production `plane(_:timestamp:)` overload flattens
    /// that and calls this helper.
    public static func planeJson(
        timestampNanos: Int64,
        id: Int,
        kind: String,
        worldPolygon: [[Float]]
    ) -> String {
        var s = ""
        s.reserveCapacity(256 + worldPolygon.count * 48)
        s.append("{")
        appendCommonHeader(&s, timestampNanos: timestampNanos, type: "plane", entity: "world/planes/\(id)")
        s.append(",\"kind\":\"")
        s.append(kind)
        s.append("\",\"polygon\":[")
        for (idx, p) in worldPolygon.enumerated() {
            if idx > 0 { s.append(",") }
            s.append("[")
            if p.count >= 3 {
                appendFloat(&s, p[0]); s.append(",")
                appendFloat(&s, p[1]); s.append(",")
                appendFloat(&s, p[2])
            }
            s.append("]")
        }
        s.append("]}\n")
        return s
    }

    // MARK: - Production overloads (ARKit types) — iOS only

    #if os(iOS)

    /// Camera pose event — derives translation and quaternion from an
    /// ARFrame's camera `transform` (`simd_float4x4`).
    public static func cameraPose(
        timestampNanos: Int64,
        camera: ARCamera,
        entity: String = "world/camera"
    ) -> String {
        let t = camera.transform
        // Translation is the 4th column of the 4x4 transform.
        let tx = t.columns.3.x
        let ty = t.columns.3.y
        let tz = t.columns.3.z
        // Rotation quaternion from the upper-left 3x3.
        let q = simd_quatf(
            simd_float3x3(
                SIMD3(t.columns.0.x, t.columns.0.y, t.columns.0.z),
                SIMD3(t.columns.1.x, t.columns.1.y, t.columns.1.z),
                SIMD3(t.columns.2.x, t.columns.2.y, t.columns.2.z)
            )
        )
        return cameraPoseJson(
            timestampNanos: timestampNanos,
            tx: tx, ty: ty, tz: tz,
            qx: q.imag.x, qy: q.imag.y, qz: q.imag.z, qw: q.real,
            entity: entity
        )
    }

    /// Plane event — derives the world-space polygon from an ARPlaneAnchor.
    public static func plane(
        timestampNanos: Int64,
        anchor: ARPlaneAnchor
    ) -> String {
        // ARKit's ARPlaneGeometry.boundaryVertices is in the anchor's local
        // space. Lift each vertex to world space via anchor.transform.
        let xform = anchor.transform
        let verts = anchor.geometry.boundaryVertices
        var worldPoly: [[Float]] = []
        worldPoly.reserveCapacity(verts.count)
        for v in verts {
            let local = SIMD4<Float>(v.x, v.y, v.z, 1.0)
            let world = xform * local
            worldPoly.append([world.x, world.y, world.z])
        }
        let kind: String
        switch anchor.alignment {
        case .horizontal: kind = "horizontal_upward"
        case .vertical: kind = "vertical"
        @unknown default: kind = "unknown"
        }
        return planeJson(
            timestampNanos: timestampNanos,
            id: anchor.identifier.hashValue,
            kind: kind,
            worldPolygon: worldPoly
        )
    }

    /// Point cloud event — derives positions from an ARFrame's raw feature
    /// points. ARKit's `ARPointCloud.points` is `[simd_float3]`, which we
    /// flatten into the `[x, y, z, x, y, z, ...]` shape our testable
    /// overload expects. Confidences are not exposed by ARKit so we emit
    /// an empty list.
    public static func pointCloud(
        timestampNanos: Int64,
        cloud: ARPointCloud,
        entity: String = "world/points"
    ) -> String {
        let pts = cloud.points
        var flat: [Float] = []
        flat.reserveCapacity(pts.count * 3)
        for p in pts {
            flat.append(p.x)
            flat.append(p.y)
            flat.append(p.z)
        }
        return pointCloudJson(
            timestampNanos: timestampNanos,
            positions: flat,
            confidences: [],
            entity: entity
        )
    }

    /// Anchor event — takes any ARAnchor (ARPlaneAnchor is handled by
    /// `plane(_:timestamp:)` instead, this overload is for user-placed
    /// AnchorEntity-backed anchors).
    public static func anchor(
        timestampNanos: Int64,
        arAnchor: ARAnchor
    ) -> String {
        let t = arAnchor.transform
        let tx = t.columns.3.x
        let ty = t.columns.3.y
        let tz = t.columns.3.z
        let q = simd_quatf(
            simd_float3x3(
                SIMD3(t.columns.0.x, t.columns.0.y, t.columns.0.z),
                SIMD3(t.columns.1.x, t.columns.1.y, t.columns.1.z),
                SIMD3(t.columns.2.x, t.columns.2.y, t.columns.2.z)
            )
        )
        return anchorJson(
            timestampNanos: timestampNanos,
            id: arAnchor.identifier.hashValue,
            tx: tx, ty: ty, tz: tz,
            qx: q.imag.x, qy: q.imag.y, qz: q.imag.z, qw: q.real
        )
    }

    #endif

    // MARK: - Helpers

    private static func appendCommonHeader(
        _ s: inout String,
        timestampNanos t: Int64,
        type: String,
        entity: String
    ) {
        s.append("\"t\":")
        s.append(String(t))
        s.append(",\"type\":\"")
        s.append(type)
        s.append("\",\"entity\":\"")
        appendEscaped(&s, entity)
        s.append("\"")
    }

    /// Writes a float as JSON number. Non-finite values (NaN/Infinity)
    /// would break the JSON line; we emit `0` instead so the line stays
    /// parseable — matching the Kotlin bridge's behaviour byte-for-byte.
    ///
    /// Finite floats are rendered by `String(describing:)` which matches
    /// Kotlin's `Float.toString()` output for the same input, so the
    /// golden-JSON tests are directly comparable across the two bridges.
    private static func appendFloat(_ s: inout String, _ f: Float) {
        if !f.isFinite {
            s.append("0")
            return
        }
        s.append(String(describing: f))
    }

    /// Minimal JSON string escaper. Entity paths may legitimately contain
    /// `/` but we escape defensively against quotes and control chars.
    private static func appendEscaped(_ s: inout String, _ str: String) {
        for c in str {
            switch c {
            case "\"": s.append("\\\"")
            case "\\": s.append("\\\\")
            case "\n": s.append("\\n")
            case "\r": s.append("\\r")
            case "\t": s.append("\\t")
            default:
                if let ascii = c.asciiValue, ascii < 0x20 {
                    s.append(String(format: "\\u%04x", ascii))
                } else {
                    s.append(c)
                }
            }
        }
    }
}
